pkg/algorithm/nodeselect/v1alpha2/select_node.go

```diff
@@ -0,0 +1,206 @@
+	nodeList, err := nodeapis.NewKubeClient().List(metav1.ListOptions{LabelSelector: getLabelSelectorString(labels)})
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get node list from the node selector")
+	}
+	if len(nodeList.Items) != 1 {
+		return "", errors.Errorf("could not get a unique node from the given node selectors")
+	}
+	if !nodeapis.NewBuilder().WithAPINode(&nodeList.Items[0]).Node.IsReady() {
+		return "", errors.Errorf("node {%s} is not ready", nodeList.Items[0].Name)
+	}
+	return nodeList.Items[0].Name, nil
+}
+
+// getLabelSelectorString returns a string of label selector form label map to be used in
+// list options.
+func getLabelSelectorString(selector map[string]string) string {
+	var selectorString string
+	for key, value := range selector {
+		selectorString = selectorString + key + "=" + value + ","
+	}
+	selectorString = selectorString[:len(selectorString)-len(",")]
+	return selectorString
+}
+
+// GetUsedNode returns a map of node for which pool has already been created.
+func (ac *Config) GetUsedNode() (map[string]bool, error) {
+	usedNode := make(map[string]bool)
+	cspList, err := csp.
+		NewKubeClient().
+		WithNamespace(ac.Namespace).
+		List(
+			metav1.
+				ListOptions{LabelSelector: string(apis.CStorPoolClusterCPK) + "=" + ac.CSPC.Name},
+		)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not list already created csp(s)")
+	}
+	for _, cspObj := range cspList.Items {
+		usedNode[cspObj.Labels[string(apis.HostNameCPK)]] = true
+	}
+	return usedNode, nil
+}
+
+// GetBDListForNode returns a list of BD from the pool spec.
+// TODO : Move it to CStorPoolCluster packgage
+func (ac *Config) GetBDListForNode(pool *apis.PoolSpec) []string {
+	var BDList []string
+	for _, group := range pool.RaidGroups {
+		for _, bd := range group.BlockDevices {
+			BDList = append(BDList, bd.BlockDeviceName)
+		}
+	}
+	return BDList
+}
+
+// ClaimBDsForNode claims a given BlockDevice for node
+// If the block device(s) is/are already claimed for any other CSPC it returns error.
+// If the block device(s) is/are already calimed for the same CSPC -- it is left as it is and can be used for
+// pool provisioning.
+// If the block device(s) is/are unclaimed, then those are claimed.
+func (ac *Config) ClaimBDsForNode(BD []string) error {
+	for _, bdName := range BD {
+		bdAPIObj, err := bd.NewKubeClient().WithNamespace(ac.Namespace).Get(bdName, metav1.GetOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "error in getting details for BD {%s} whether it is claimed", bdName)
+		}
+		if bd.BuilderForAPIObject(bdAPIObj).BlockDevice.IsClaimed() {
+			IsClaimedBDUsable, err := ac.IsClaimedBDUsable(bdAPIObj)
+			if err != nil {
+				return errors.Wrapf(err, "error in getting details for BD {%s} for usability", bdName)
+			}
+			if !IsClaimedBDUsable {
+				return errors.Errorf("BD {%s} already in use", bdName)
+			}
+		}
+
+		err = ac.ClaimBD(bdAPIObj)
```

Comment:
also, we should not create if BD is still unclaimed, as there can be case where BDC is created, but, its claim is not fulfilled yet

Reply:
Will fix it

Reply:
done